File: utilities/dates/go/date_utils.go
package dates

import "time"

func ParseDate(value string) (*time.Time, error) {
	if value == "" { return nil, nil }
	if parsed, err := time.Parse(time.RFC3339, value); err == nil { return &parsed, nil }
	parsed, err := time.ParseInLocation("2006-01-02 15:04:05", value, time.UTC)
	if err != nil { return nil, err }
	return &parsed, nil
}

func DaysBetween(start, end string) (*int64, error) {
	startDate, err := ParseDate(start)
	if err != nil { return nil, err }
	endDate, err := ParseDate(end)
	if err != nil { return nil, err }
	if startDate == nil || endDate == nil { return nil, nil }
	days := int64(endDate.Sub(*startDate) / (24 * time.Hour))
	return &days, nil
}

File: utilities/dates/csharp/DateUtils.cs
using System.Globalization;

namespace PolyglotUtilities.Dates;

public static class DateUtils
{
    public static DateTimeOffset? ParseDate(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return null;
        }

        return DateTimeOffset.TryParse(
            value.Trim(),
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
            out var result)
            ? result
            : null;
    }

    public static long? DaysBetween(string? start, string? end)
    {
        var startDate = ParseDate(start);
        var endDate = ParseDate(end);
        if (startDate is null || endDate is null)
        {
            return null;
        }

        return (long)Math.Floor(
            (endDate.Value - startDate.Value).TotalSeconds / 86400.0 + 0.5);
    }
}
